Feat/testimonial form/create and edit form testimonials

andreajhl opened the pull request:
-Add get, post and put functions in TestimonialsServer.
-Add activity creation and editing form, made with formik and ckeditor.
-Add modular function to render Formik form errors
-Add modular function to get url of file type entries
-You don't have styles yet because you have to decide as a group.

Form:

-title: validation => long (minimum 4 letters)
-image: validation => must exist and aceept jpg or png, render image over input on loading, as expected
-description: validation => must exist

![testimonials](https://user-images.githubusercontent.com/75461111/139378538-f22cdd9a-b0c7-4629-92e6-c8ffb86e939c.png)

Commit:
feat: :sparkles: add form for edit and create testimonials

it extracts in two functions the rendering of errors, and the obtaining of the url of the image inout

Commit:
feat: :sparkles: add request HTTP ( get, post, put) for the form testimonial

Commit:
feat: :sparkles: add modularized functions to get the url of an image

add modularized functions to get the url of an image, and save it in a state

Commit:
feat: :sparkles: adds modularized functions to render form errors

adds modularized functions to render form errors made with formik, requires name and error

Commit:
feat: :sparkles: function to validate form the testimonials forms

function to validate form forms, for name, description and image fields, returning the corresponding errors

Commit:
feat: :sparkles: form modularization

removes the functionality of replace validate errors to other components in the common folder

saralavanchy commented on src/Components/Testimonials/TestimonialsForm.js:
```diff
@@ -1,32 +1,96 @@
-       description: '' 
-    });
+  const [imgState,setImgState] = useState();
+  const [dataForm, setDataForm] = useState({
```
the name of this should be more representative of what it is, maybe just testimonial its a good candidate
```
const [testimonial, setTestimonial] = useState({...code here...});
```

saralavanchy commented on src/Components/Testimonials/TestimonialsForm.js:
```diff
@@ -1,32 +1,96 @@
-        e.preventDefault();
-        console.log(initialValues);
-    }
+  const handleChaneCkEditor = (editor, setFieldValue) => {
```
handleChangeDescription maybe its a better name. Its using CkEditor but its changing the description and that's the main functionallity of this function.

saralavanchy commented on src/Components/Testimonials/TestimonialsForm.js:
```diff
@@ -1,32 +1,96 @@
-        console.log(initialValues);
-    }
+  const handleChaneCkEditor = (editor, setFieldValue) => {
+    setFieldValue("description", editor.getData())
```
in this case this parameter name can be setDescription and its clearest what's doing

saralavanchy commented on src/Components/Testimonials/TestimonialsForm.js:
```diff
@@ -1,32 +1,96 @@
-        <form className="form-container" onSubmit={handleSubmit}>
-            <input className="input-field" type="text" name="name" value={initialValues.name} onChange={handleChange} placeholder="Testimonial Title"></input>
-            <input className="input-field" type="text" name="description" value={initialValues.description} onChange={handleChange} placeholder="Testimonial description"></input>
+  const handleChangeinputFile = (e, setFieldValue) =>{ 
```
in this case a better name maybe its 
```
const handleChangeImage = (e, setImage) =>
```
always think on whats the main goal of the function and what can be done to make clearest the code for a programmer that never read it. with current name I had to go down and read on the ui where its being used to be sure of what it do. With a name like handleChangeImage there are no dudes of that it change the image only by reading the code. That's the meaning of that the code should be enought documentation for itself.

saralavanchy commented on src/Components/Testimonials/TestimonialsForm.js:
```diff
@@ -1,32 +1,96 @@
+          ...dataForm
+        }}
+        validate={(values)=>validateFormTestimonials(values)}
+        onSubmit={(values)=>{
```
this onSubmit functionallity should go on  a separated function,

saralavanchy commented on src/Components/Testimonials/TestimonialsForm.js:
```diff
@@ -1,32 +1,96 @@
+import { CKEditor } from '@ckeditor/ckeditor5-react';
+import { UrlInput } from '../common/getUrlInputFile';
+import ClassicEditor from '@ckeditor/ckeditor5-build-classic';
+import { validateFormTestimonials, replaceCKeditor } from '../common/validateFormTestimonials';
```
if it goes on common it should go on a validations file because validateFormTestimonials is not too common, its especific

saralavanchy commented on src/Components/common/getUrlInputFile.js:
```diff
@@ -0,0 +1,12 @@
+
+export const UrlInput = (e,setUrl,setFieldValue) =>{
```
I think that this have more than one responsability but im not sure because its not clear what is being done here.

saralavanchy commented on src/Components/common/getUrlInputFile.js:
```diff
@@ -0,0 +1,12 @@
+
```
the name of the file should be just image because with the current name you are limiting the use of this file just to the current function and its on a common file, so, all the common functionallities for a image should be here

andreajhl replied:
I do not understand what we refer to with just an image?

saralavanchy commented on src/Components/common/getUrlInputFile.js:
```diff
@@ -0,0 +1,12 @@
+
+export const UrlInput = (e,setUrl,setFieldValue) =>{
+
+  const imageFile = e.target.files[0];
```
this imageFile should come from the parameters because its specific of the TestimonialForm

saralavanchy commented on src/Components/common/getUrlInputFile.js:
```diff
@@ -0,0 +1,12 @@
+
+  const imageFile = e.target.files[0];
+  const imageUrl = new FileReader();
+  imageUrl.readAsDataURL(imageFile)
```
this is one responsability read the file. Maybe it can be a function called readImageFile that returns this line.
```
const readImageFile = (imageFile, onLoadCallback) => {
      const imageUrl = new FileReader();
      imageUrl.readAsDataURL(imageFile);
      imageUrl.onload = (event) => onLoadCallback(event);
}
```

andreajhl replied:
I'm sorry, I don't understand what you mean by this, or what do you mean with the responsibility of the form?

saralavanchy commented on src/Components/common/getUrlInputFile.js:
```diff
@@ -0,0 +1,12 @@
+  const imageUrl = new FileReader();
+  imageUrl.readAsDataURL(imageFile)
+  imageUrl.onload=(e)=>{
+    setFieldValue("image",e.target.result)
```
this is the second responsability that is to setTheValue and its responsability of the TestimonialForm

andreajhl replied:
I'm sorry, I don't understand what you mean by this, or what do you mean with the responsibility of the form?

saralavanchy commented on src/Components/common/getUrlInputFile.js:
```diff
@@ -0,0 +1,12 @@
+  imageUrl.readAsDataURL(imageFile)
+  imageUrl.onload=(e)=>{
+    setFieldValue("image",e.target.result)
+    setUrl(e.target.result)
```
this is the third responsability that it..., well, I do not know what it do. It sets somewhere the url of the result of an event. But it looks to be responsability of the TestimonialForm

andreajhl replied:
I'm sorry, I don't understand what you mean by this, or what do you mean with the responsibility of the form?

saralavanchy commented on src/Components/common/messagesForm.js:
```diff
@@ -0,0 +1,11 @@
+
```
this is not a messageForm, its an CustomErrorMessage, so, it should be the name of the file and of the component

saralavanchy commented on src/Components/common/validateFormTestimonials.js:
```diff
@@ -0,0 +1,18 @@
+export const validateFormTestimonials = (values) => {
```
this is a validations file, so the name of the file should be just validations. Then inside the validations files you can have a specific formTestimonialValidations

saralavanchy commented on src/Components/common/validateFormTestimonials.js:
```diff
@@ -0,0 +1,18 @@
+    if(!values.image) errors.image='You need to add a image';
+
+    if(!values.name) errors.name='You need to enter a name';
+    if(values.name.length<4) errors.name='The name must have more than 4 letters';
```
this validation should be a separated function in string or list common file.
```
const isSmallerThan = (list, maxSize) => list.length < maxSize;
```
then you use it here
```
...
if(isSmallerThan(values.name, 4) .....
```

saralavanchy commented on src/Components/common/validateFormTestimonials.js:
```diff
@@ -0,0 +1,18 @@
+
+}
+
+export const replaceCKeditor = (values)=>{
```
this is not a validation function so it shoudl be on  ckEditor file. Also, im pretty sure that it do not replaces the ckEditor for something, it setDescriptionValue. And, because description its specific of the testimonial it should go on TestimonialForm file. If you can do it a little more generic, it should be on a ckEditor file.

Something more generic could be something like
```
export const setCKEditorText = (values, field) => {
    values[field] = values[field].replace(/'<p>'|['</p>']/gi,'');
    return values;
}
```

saralavanchy commented on src/Services/testimonialService.js:
```diff
@@ -0,0 +1,20 @@
+    return data;
+};
+
+export const modifiedTestimonial = async (id,body) => {
```
modify, the name of the functions should be on infinitive

saralavanchy commented on src/Services/testimonialService.js:
```diff
@@ -0,0 +1,20 @@
+export const createTestimonial = async (body) => {
+    console.log(body)
+    let {data}= await axios.post(`${URL}/testimonials`,body);
+    console.log(data)
```
this and the previous console.log are not necessary

saralavanchy commented on src/Components/Testimonials/TestimonialsForm.js:
```diff
@@ -1,32 +1,96 @@
+        validate={(values)=>validateFormTestimonials(values)}
+        onSubmit={(values)=>{
+          let valuesForm=replaceCKeditor(values)
+          testimonialId ?   modifiedTestimonial( testimonialId, valuesForm ) : createTestimonial(valuesForm)
```
the logic to create or modify the testimonial its a bussiness logic so it should be on the testimonialService. There should be a function something like:
```
const createOrUpdateTestimonial = (testimonialId, valuesForm) => {
    testimonialId ?   modifiedTestimonial( testimonialId, valuesForm ) : createTestimonial(valuesForm)
}
```

saralavanchy commented on src/Components/Testimonials/TestimonialsForm.js:
```diff
@@ -1,32 +1,96 @@
+        }}
+        validate={(values)=>validateFormTestimonials(values)}
+        onSubmit={(values)=>{
+          let valuesForm=replaceCKeditor(values)
```
this is updatedValues instead valuesForm because both values and valuesForm are form values, the diference is that one is as it is and the second one is updated or cleaned value. So the name should be updatedValues or cleanedValues

saralavanchy commented on src/Components/Testimonials/TestimonialsForm.js:
```diff
@@ -1,32 +1,96 @@
+        initialValues={{
+          ...dataForm
+        }}
+        validate={(values)=>validateFormTestimonials(values)}
```
validateTestimonialsForm its a better constructed than validateFormTestimonials

Commit:
feat: :fire: remove and delete de function in folder common

Commit:
feat: :sparkles: add function setCKEditorText

replace "<p></p>" in the text and set value in the state

Commit:
feat: :sparkles: adds validation folder, includes validations for the testimonial form and function to validate the length of the name or title

Commit:
feat: :sparkles: add function to render errors in the form

Commit:
feat: :sparkles: adds function to dispatch the corresponding request

Commit:
feat: :sparkles: modify the name of some variables and element

Commit:
feat: :fire: delete file

Commit:
feat: :seedling: modified route to import file

modified route to import file "setCKEditor" and "validations"

saralavanchy commented on src/Services/testimonialService.js:
```diff
@@ -0,0 +1,22 @@
+    return data;
+};
+
+export const dispatchRequestcreateOrUpdateTestimonial = (testimonialId,body)=>{
```
dispatchRequestcreateOrUpdateTestimonial  es muy dificil de leer, createOrUpdateTestimonial es simple y representativo de lo que hace

Commit:
feat: :fire: delete deprecated file

Commit:
feat: :sparkles: add file file with functions to read and set the value of input file

add the seturl function that invokes the function that reads the url and sets it through a callback function

Commit:
feat: :speech_balloon: modified name function "validateTestimonial => validateTestimonialForm"

Commit:
feat: :speech_balloon: modify the names of some functions

modify the names of some functions to make them more understandable and representative

Commit:
feat: :speech_balloon: modify the names of some functions to make them more understandable and representative

andreajhl committed:
Merge branch 'development' into feat/testimonial-form/create-and-edit-form-testimonials

Commit:
feat: :sparkles: add route in file App for render edit testimonial

Commit:
feat: :sparkles: validation to decide which request to make

Commit:
feat: :sparkles: modifies the initial value of the form momentarily, until you have redux

Commit:
feat: :fire: delete code in App

Commit:
feat: :fire: delete unused files

Commit:
feat: :fire: remove import to counter slider

andreajhl committed:
Merge branch 'development' into feat/testimonial-form/create-and-edit-form-testimonials

andreajhl committed:
Merge branch 'development' into feat/testimonial-form/create-and-edit-form-testimonials

andreajhl committed:
Merge branch 'development' into feat/testimonial-form/create-and-edit-form-testimonials
